conjure-java-jackson-serialization/src/main/java/com/palantir/conjure/java/serialization/CaffeineCachingTypeFactory.java

```diff
@@ -0,0 +1,145 @@
+import javax.annotation.Nullable;
+
+final class CaffeineCachingTypeFactory extends TypeFactory {
+    private static final SafeLogger log = SafeLoggerFactory.get(CaffeineCachingTypeFactory.class);
+
+    CaffeineCachingTypeFactory() {
+        super(new CaffeineLookupCache());
+    }
+
+    CaffeineCachingTypeFactory(
+            CaffeineLookupCache typeCache,
+            TypeParser parser,
+            @Nullable TypeModifier[] modifiers,
+            ClassLoader classLoader) {
+        super(typeCache, parser, modifiers, classLoader);
+    }
+
+    private static CaffeineLookupCache asCaffeine(LookupCache<Object, JavaType> typeCache) {
+        if (typeCache instanceof CaffeineLookupCache) {
+            return (CaffeineLookupCache) typeCache;
+        } else if (typeCache == null || typeCache.size() == 0) {
+            // Use a caffeine cache instead of the provided empty cache
+            return new CaffeineLookupCache();
+        } else if (typeCache instanceof LRUMap) {
+            CaffeineLookupCache cache = new CaffeineLookupCache();
+            LRUMap<Object, JavaType> defaultJacksonCacheType = (LRUMap<Object, JavaType>) typeCache;
+            defaultJacksonCacheType.contents(cache::put);
+            return cache;
+        } else {
+            log.error(
+                    "Unknown LookupCache, using a new caffeine cache instead",
+                    SafeArg.of("cacheType", typeCache.getClass()));
+            return new CaffeineLookupCache();
+        }
+    }
+
+    @Override
+    public CaffeineCachingTypeFactory withModifier(TypeModifier mod) {
+        // Updating modifiers clears the cache
+        return new CaffeineCachingTypeFactory(
+                new CaffeineLookupCache(), _parser, computeModifiers(_modifiers, mod), _classLoader);
+    }
+
+    @Nullable
+    private static TypeModifier[] computeModifiers(TypeModifier[] existing, TypeModifier newModifier) {
+        // Semantics are based on the jackson-databind `withModifier` implementation.
+        if (newModifier == null) {
+            return null;
+        }
+        if (existing == null || existing.length == 0) {
+            return new TypeModifier[] {newModifier};
+        }
+        return ArrayBuilders.insertInListNoDup(existing, newModifier);
+    }
+
+    @Override
+    public CaffeineCachingTypeFactory withClassLoader(ClassLoader classLoader) {
+        return new CaffeineCachingTypeFactory(asCaffeine(_typeCache), _parser, _modifiers, classLoader);
+    }
+
+    @Override
+    public CaffeineCachingTypeFactory withCache(LRUMap<Object, JavaType> cache) {
+        return withCache((LookupCache<Object, JavaType>) cache);
+    }
+
+    @Override
+    public CaffeineCachingTypeFactory withCache(LookupCache<Object, JavaType> cache) {
+        return new CaffeineCachingTypeFactory(asCaffeine(cache), _parser, _modifiers, _classLoader);
+    }
+
+    private static final class CaffeineLookupCache implements LookupCache<Object, JavaType> {
```

Comment:
Initially I opted to implement only the cache component, however we must extend `TypeFactory` so that calls to `withModifier(modifier)`don't drop our cache in favor of the default implementation.